File: netferry-desktop/src-tauri/src/crypto.rs
use aes_gcm::aead::{Aead, KeyInit};
use aes_gcm::{Aes256Gcm, Nonce};
use base64::engine::general_purpose::STANDARD as B64;
use base64::Engine;
use rand::RngCore;

fn export_key() -> Result<[u8; 32], String> {
    let hex = option_env!("NETFERRY_EXPORT_KEY")
        .ok_or("Profile export is not available in this build")?;
    let hex = hex.trim();
    if hex.len() != 64 {
        return Err("NETFERRY_EXPORT_KEY must be 64 hex characters (32 bytes)".into());
    }
    let mut key = [0u8; 32];
    for (i, chunk) in hex.as_bytes().chunks(2).enumerate() {
        key[i] = u8::from_str_radix(std::str::from_utf8(chunk).unwrap(), 16)
            .map_err(|_| "NETFERRY_EXPORT_KEY contains invalid hex")?;
    }
    Ok(key)
}

/// Encrypt plaintext bytes with AES-256-GCM.
/// Returns base64-encoded string: nonce (12 bytes) || ciphertext+tag.
pub fn encrypt(plaintext: &[u8]) -> Result<String, String> {
    let key = export_key()?;
    let cipher = Aes256Gcm::new_from_slice(&key).map_err(|e| e.to_string())?;
    let mut nonce_bytes = [0u8; 12];
    rand::thread_rng().fill_bytes(&mut nonce_bytes);
    let nonce = Nonce::from_slice(&nonce_bytes);
    let ciphertext = cipher
        .encrypt(nonce, plaintext)
        .map_err(|e| format!("Encryption failed: {e}"))?;
    let mut combined = nonce_bytes.to_vec();
    combined.extend_from_slice(&ciphertext);
    Ok(B64.encode(&combined))
}

/// Decrypt a base64-encoded string produced by `encrypt`.
pub fn decrypt(encoded: &str) -> Result<Vec<u8>, String> {
    let key = export_key()?;
    let combined = B64
        .decode(encoded.trim())
        .map_err(|e| format!("Invalid base64: {e}"))?;
    if combined.len() < 13 {
        return Err("Encrypted data too short".into());
    }
    let (nonce_bytes, ciphertext) = combined.split_at(12);
    let cipher = Aes256Gcm::new_from_slice(&key).map_err(|e| e.to_string())?;
    let nonce = Nonce::from_slice(nonce_bytes);
    cipher
        .decrypt(nonce, ciphertext)
        .map_err(|_| "Decryption failed — wrong key or corrupted data".into())
}

File: netferry-relay/internal/profile/crypto.go
package profile

import (
	"crypto/aes"
	"crypto/cipher"
	"encoding/base64"
	"encoding/hex"
	"fmt"
	"strings"
)

// ExportKey is the AES-256 key (hex-encoded, 64 chars) used to decrypt
// exported .nfprofile files. Must match the desktop NETFERRY_EXPORT_KEY.
// Set at build time via ldflags:
//
//	go build -ldflags="-X github.com/hoveychen/netferry/relay/internal/profile.ExportKey=<hex>" ...
var ExportKey = ""

func exportKeyBytes() ([]byte, error) {
	if ExportKey == "" {
		return nil, fmt.Errorf("profile export is not available in this build (NETFERRY_EXPORT_KEY not baked in)")
	}
	key, err := hex.DecodeString(ExportKey)
	if err != nil {
		return nil, fmt.Errorf("invalid export key: %w", err)
	}
	if len(key) != 32 {
		return nil, fmt.Errorf("export key must be 32 bytes (64 hex chars)")
	}
	return key, nil
}

// Decrypt takes a base64-encoded ciphertext (nonce || ciphertext+tag) produced
// by the desktop `encrypt` function and returns the plaintext bytes.
func Decrypt(encrypted string) ([]byte, error) {
	key, err := exportKeyBytes()
	if err != nil {
		return nil, err
	}

	combined, err := base64.StdEncoding.DecodeString(strings.TrimSpace(encrypted))
	if err != nil {
		return nil, fmt.Errorf("invalid base64: %w", err)
	}
	if len(combined) < 13 {
		return nil, fmt.Errorf("encrypted data too short")
	}

	nonce := combined[:12]
	ciphertext := combined[12:]

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}

	plaintext, err := gcm.Open(nil, nonce, ciphertext, nil)
	if err != nil {
		return nil, fmt.Errorf("decryption failed — wrong key or corrupted data")
	}
	return plaintext, nil
}
